File: services/notification-service/src/routes/health.py
"""
Health Check Routes
"""

from flask import Blueprint, jsonify
from datetime import datetime
import time

health_bp = Blueprint('health', __name__)

START_TIME = time.time()

@health_bp.route('/health')
def health():
    """Liveness probe"""
    return jsonify({
        'status': 'healthy',
        'service': 'notification-service',
        'timestamp': datetime.utcnow().isoformat(),
        'uptime': int(time.time() - START_TIME)
    })

@health_bp.route('/ready')
def ready():
    """Readiness probe"""
    from src.services.sns_service import SNSService
    
    try:
        sns_service = SNSService()
        sns_service.check_connection()
        sns_status = 'up'
    except Exception as e:
        sns_status = f'down: {str(e)}'
    
    is_ready = sns_status == 'up'
    
    return jsonify({
        'status': 'ready' if is_ready else 'not ready',
        'service': 'notification-service',
        'timestamp': datetime.utcnow().isoformat(),
        'dependencies': {
            'sns': sns_status
        }
    }), 200 if is_ready else 503

File: services/api-gateway/src/routes/health.js
/**
 * Health Check Routes
 */

const express = require('express');
const router = express.Router();

const startTime = Date.now();

/**
 * Liveness probe - is the service running?
 */
router.get('/health', (req, res) => {
    res.json({
        status: 'healthy',
        service: 'api-gateway',
        timestamp: new Date().toISOString(),
        uptime: Math.floor((Date.now() - startTime) / 1000)
    });
});

/**
 * Readiness probe - is the service ready to accept traffic?
 */
router.get('/ready', async (req, res) => {
    const checks = {
        auth: await checkService(process.env.AUTH_SERVICE_URL),
        user: await checkService(process.env.USER_SERVICE_URL),
        order: await checkService(process.env.ORDER_SERVICE_URL),
        notification: await checkService(process.env.NOTIFICATION_SERVICE_URL)
    };

    const allHealthy = Object.values(checks).every(c => c.status === 'up');

    res.status(allHealthy ? 200 : 503).json({
        status: allHealthy ? 'ready' : 'degraded',
        service: 'api-gateway',
        timestamp: new Date().toISOString(),
        dependencies: checks
    });
});

/**
 * Check if a downstream service is healthy
 */
async function checkService(url) {
    if (!url) {
        return { status: 'unknown', message: 'URL not configured' };
    }

    try {
        const axios = require('axios');
        const response = await axios.get(`${url}/health`, { timeout: 2000 });
        return { status: 'up', latency: response.headers['x-response-time'] };
    } catch (error) {
        return { status: 'down', error: error.message };
    }
}

module.exports = router;
